shell/renderer/api/atom_api_renderer_ipc.cc

```diff
@@ -69,45 +70,71 @@ class IPCRenderer : public gin::Wrappable<IPCRenderer> {
   const char* GetTypeName() override { return "IPCRenderer"; }
 
  private:
-  void Send(bool internal,
+  void Send(v8::Isolate* isolate,
+            bool internal,
             const std::string& channel,
-            const base::ListValue& arguments) {
-    electron_browser_ptr_->get()->Message(internal, channel, arguments.Clone());
+            v8::Local<v8::Value> arguments) {
+    blink::CloneableMessage message;
+    if (!mate::ConvertFromV8(isolate, arguments, &message)) {
+      return;
+    }
+    electron_browser_ptr_->get()->Message(internal, channel,
+                                          std::move(message));
   }
 
   v8::Local<v8::Promise> Invoke(v8::Isolate* isolate,
                                 bool internal,
                                 const std::string& channel,
-                                const base::Value& arguments) {
-    electron::util::Promise<base::Value> p(isolate);
+                                v8::Local<v8::Value> arguments) {
+    blink::CloneableMessage message;
+    if (!mate::ConvertFromV8(isolate, arguments, &message)) {
+      return v8::Local<v8::Promise>();
```

Comment:
Is this promise rejected or does the method throw an error and return an empty promise?

Reply:
When the serialization fails, the method throws an error. Also, this isn't a `Promise` value, it's an empty `Local`.

Reply:
My concern here was is that error still catchable with .catch?

If it's an empty promise and the method throws then .catch won't work which is incompatible with a promise based API

Reply:
yes, it's still catchable.

```js
ipcRenderer.invoke('foo', ()=>{}).catch(e => console.log('bloop'))
// => bloop
```